src/schemas/ProgramDonorPublishedAnalysisByDateRange/resolvers/programDonorPublishedAnalysisByDateRange.ts

```diff
@@ -0,0 +1,124 @@
+import {
+  BaseQueryArguments,
+  DonorFields,
+  EsAggs,
+  ProgramDonorGqlResponse,
+} from './types';
+import { Client } from '@elastic/elasticsearch';
+import { ELASTICSEARCH_PROGRAM_DONOR_DASHBOARD_INDEX } from 'config';
+import { UserInputError } from 'apollo-server-express';
+import { convertStringToISODate, validateISODate } from 'utils/dateUtils';
+import { ELASTICSEARCH_DATE_TIME_FORMAT } from '../../../constants/elasticsearch';
+import { differenceInDays, sub as subDate, formatISO } from 'date-fns';
+import logger from 'utils/logger';
+
+const ERROR_TITLE = 'ProgramDonorPublishedAnalysisByDateRange:'
+
+const programDonorPublishedAnalysisByDateRangeResolver: (
+  esClient: Client,
+) => GraphQLFieldResolver<
+  unknown,
+  GlobalGqlContext,
+  BaseQueryArguments & {
+    bucketCount: number;
+    dateRangeFrom: string;
+    dateRangeTo: string;
+    donorFields: DonorFields[];
+  }
+> = esClient => async (source, args, context): Promise<ProgramDonorGqlResponse[]> => {
+  const {
+    bucketCount,
+    dateRangeTo,
+    dateRangeFrom,
+    donorFields,
+    programShortName
+  } = args;
+
+  const areDatesValid = validateISODate(dateRangeFrom) && validateISODate(dateRangeTo);
+  if (!areDatesValid) {
+    throw new UserInputError(`${ERROR_TITLE} Dates must be in ISO format`, {
+      dateRangeFrom,
+      dateRangeTo
+    });
+  }
+
+  const isoDateRangeFrom = convertStringToISODate(dateRangeFrom);
+  const isoDateRangeTo = convertStringToISODate(dateRangeTo);
+
+  const daysInRange = differenceInDays(isoDateRangeTo, isoDateRangeFrom);
+  if (daysInRange < bucketCount) {
+    throw new UserInputError(`${ERROR_TITLE
+      } Days in range must be greater than or equal to bucket count`, {
+      bucketCount,
+      dateRangeFrom,
+      dateRangeTo,
+      daysInRange
+    });
+  }
+
+  const bucketDates = [...Array(bucketCount).keys()]
+    .sort((a, b) => b - a)
+    .map((bucketIndex: number) => subDate(isoDateRangeTo, 
+      { days: Math.floor(daysInRange / bucketCount * bucketIndex) }
+    ))
+    .map((bucketDate: Date) => formatISO(bucketDate));
+
+  const esQuery = esb
+    .requestBodySearch()
+    .size(0)
+    .query(
+      esb.boolQuery()
+        .filter(esb.termQuery('programId', programShortName))
+        .should(donorFields
+          .map((donorField: DonorFields) => esb.existsQuery(donorField)))
+        .minimumShouldMatch(1)
+    )
+    .aggs(donorFields.map((donorField: DonorFields) => esb
+      .dateRangeAggregation(donorField, donorField)
+      .format(ELASTICSEARCH_DATE_TIME_FORMAT)
+      .ranges(bucketDates.map(bucketDate => ({ to: bucketDate })))
+    ));
+
+  const esAggs: EsAggs = await esClient
+    .search({ 
+      body: esQuery,
+      index: ELASTICSEARCH_PROGRAM_DONOR_DASHBOARD_INDEX,
+    })
+    .then(res => res.body.aggregations)
+    .catch(err => {
+      logger.error(`${ERROR_TITLE} Error reading data from Elasticsearch: `, err);
+      return {} as EsAggs;
```

Comment:
Why are we returning no data instead of throwing an error for GQL to communicate? We would want to show "error fetching data" not "no data"

Reply:
(from slack convo) because that's how we did it somewhere else, but it's not good form.

updated to return empty data and error object